# Review and Helpfulness Analysis


In [1]:
# Importing Libraries
import pandas as pd   # For data manipulation and analysis
import numpy as np    # For numerical operations

In [2]:
df=pd.read_csv('Reviews.csv')   # Reading the csv file and assigning it to  variable df
df.head()                       # Using pandas function head() for displaying first 5 rows of data  

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   
3                     3                       3      2  1307923200   
4                     0                       0      5  1350777600   

                 Summary                                               Text  
0  Good Quality Dog Food  I have bought several of the Vitality canned d...  
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...  
2  "Delight" says it all  This is a confection that has been around a fe...  
3         Cough Medicine  If you are looking for the secret ingredient i...  
4            Great taffy  Great taffy at a great price.  There was a wid...

In [3]:
df.info() #Using pandas info() function to disply the structure of the file and content's with there data types

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 568454 entries, 0 to 568453
Data columns (total 10 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Id                      568454 non-null  int64 
 1   ProductId               568454 non-null  object
 2   UserId                  568454 non-null  object
 3   ProfileName             568428 non-null  object
 4   HelpfulnessNumerator    568454 non-null  int64 
 5   HelpfulnessDenominator  568454 non-null  int64 
 6   Score                   568454 non-null  int64 
 7   Time                    568454 non-null  int64 
 8   Summary                 568427 non-null  object
 9   Text                    568454 non-null  object
dtypes: int64(5), object(5)
memory usage: 43.4+ MB


In [4]:
df.describe() # Using pandas describe() function to display descriptive statistics excluding the NaN values

Id  HelpfulnessNumerator  HelpfulnessDenominator  \
count  568454.000000         568454.000000            568454.00000   
mean   284227.500000              1.743817                 2.22881   
std    164098.679298              7.636513                 8.28974   
min         1.000000              0.000000                 0.00000   
25%    142114.250000              0.000000                 0.00000   
50%    284227.500000              0.000000                 1.00000   
75%    426340.750000              2.000000                 2.00000   
max    568454.000000            866.000000               923.00000   

               Score          Time  
count  568454.000000  5.684540e+05  
mean        4.183199  1.296257e+09  
std         1.310436  4.804331e+07  
min         1.000000  9.393408e+08  
25%         4.000000  1.271290e+09  
50%         5.000000  1.311120e+09  
75%         5.000000  1.332720e+09  
max         5.000000  1.351210e+09

In [5]:
df.columns.to_list() # Using Columns().to_list() to display the column names in list format

['Id',
 'ProductId',
 'UserId',
 'ProfileName',
 'HelpfulnessNumerator',
 'HelpfulnessDenominator',
 'Score',
 'Time',
 'Summary',
 'Text']

## 1 Review scores are generally positive:
## What is the distribution of review scores? Analyze the data using Pandas.

In [6]:
print(df['Score'].describe())

count    568454.000000
mean          4.183199
std           1.310436
min           1.000000
25%           4.000000
50%           5.000000
75%           5.000000
max           5.000000
Name: Score, dtype: float64


### By this we can conclude
### The total count of reviews are 568454.000000
### The Average of the reviews is 4.183199
### The 25% of the reviews are 4.000000
### The 50% of the reviews are 5.000000
### The 75% of the reviews are 5.000000
### The minimum of the review is 1.000000 and maximum is 5.000000

--------------------------------------------------------------------------------------------------------------------------------

## 2 Helpful reviews are often marked as such by many users</b>:
## How does the helpfulness numerator compare to the helpfulness denominator across reviews? Perform a statistical analysis using NumPy</b>

In [7]:
helpfulness_narr=np.array([df['HelpfulnessNumerator']])      # Using Numpy and assigning the values of HelpfulnessNumerator  to helpfulness_narr
helpfulness_darr=np.array([df['HelpfulnessDenominator']])    # Using Numpy and assigning the values of HelpfulnessDenominator  to helpfulness_darr

In [8]:
print(np.mean(helpfulness_narr))   # Using mean() function in numpy finding the average of helpfulnessnumerator 
print(np.mean(helpfulness_darr))   # Using mean() function in numpy finding the average of helpfulnessdenominator 

1.7438174416927315
2.2288100708236724


In [9]:
print(np.std(helpfulness_narr)) 
print(np.std(helpfulness_darr))

7.636505989904315
8.28973300170383


In [10]:
helpfulness_ration=np.mean(helpfulness_narr)/np.mean(helpfulness_darr) # Finding the ration between mean values of numerator and denominator 
helpfulness_ration*100      #converting the value to percentage ~

78.23984037582402

### This means that on average, about 78.26% of the users who rated the helpfulness of the reviews found them helpful, indicating a generally positive perception of the reviews.

### 3 There are differences in review scores among different products:
### What are the average review scores for different products? Analyze the data using Pandas.

In [11]:
average_review_score=df.groupby('ProductId')['Score'].mean().reset_index() # Grouping by ProductId  and finding the average of score 

In [12]:
average_review_score.columns=['Product_Id','Average_Review_Score'] # Renaming the column names of DateFrame

### This is Average Score of each product based on ProductId

In [13]:
average_review_score

Product_Id  Average_Review_Score
0      0006641040              4.351351
1      141278509X              5.000000
2      2734888454              3.500000
3      2841233731              5.000000
4      7310172001              4.751445
...           ...                   ...
74253  B009UOFTUI              1.000000
74254  B009UOFU20              1.000000
74255  B009UUS05I              5.000000
74256  B009WSNWC4              5.000000
74257  B009WVB40S              5.000000

[74258 rows x 2 columns]

### 4 Length of the review text correlates with the review score:
### Is there a correlation between the length of the review text and the review score? Perform a correlation analysis using Pandas

In [14]:
df['Len_Text']=df['Text'].apply(len) #applying length fucntion on each Text review and storing them in DateFrame

In [15]:
df['Len_Text'] 

0         263
1         190
2         509
3         219
4         140
         ... 
568449    149
568450    252
568451    372
568452    200
568453    111
Name: Len_Text, Length: 568454, dtype: int64

In [16]:
df['Score']

0         5
1         1
2         4
3         2
4         5
         ..
568449    5
568450    2
568451    5
568452    5
568453    5
Name: Score, Length: 568454, dtype: int64

In [17]:
correlation = df['Len_Text'].corr(df['Score']) #using corr Finding the Correlation between length of the Review Text and Score

In [18]:
correlation

-0.07733381425483786

#### By performing a correlation analysis, we found that the correlation coefficient between the length of the review text and the review score is-0.0773. This indicates a very weak negative correlation, suggesting that the length of a review text has little to no significant impact on the review score. Therefore, other factors may play a more important role in influencing the review scores.

### 5 Certain users are more active in leaving reviews:
### Who are the top 10 most active reviewers, and how many reviews have they left? Analyze the data using Pandas.

In [19]:
review_counts = df.groupby('UserId').size().reset_index(name='review_count')
# Grouping By userId size() function to number of reviews and reset index() to convert back series to DataFram

In [20]:
review_counts.max()

UserId          AZZZOVIBXHGDR
review_count              448
dtype: object

In [21]:
top_review_counts=review_counts.sort_values(by='review_count', ascending=False).head(10) 
#Sorting the values in descending order and head to print top 10

#### Based on the analysis These are the Top 10 UserId with most number of Reviews

In [22]:
top_review_counts

UserId  review_count
182011  A3OXHLG6DIBRW8           448
65586   A1YUL9PCJR3JTY           421
252275   AY12DBB0U420B           389
82919   A281NPSIMI1C2R           365
66162   A1Z54EM24Y40LL           256
55736   A1TMAVN4CEM8U8           204
110863  A2MUGFV2TDQ47K           201
191283  A3TVZM3ZIXG8YW           199
183161  A3PJZ8TU8FDQ1K           178
238453   AQQLWCMRNDFGI           176

### 6 Reviews have become more detailed over time:
### How has the average length of reviews changed over time? Perform a time series analysis using Pandas.

In [23]:
df['review_date'] = pd.to_datetime(df['Time'])

In [24]:
df['review_date']

0        1970-01-01 00:00:01.303862400
1        1970-01-01 00:00:01.346976000
2        1970-01-01 00:00:01.219017600
3        1970-01-01 00:00:01.307923200
4        1970-01-01 00:00:01.350777600
                      ...             
568449   1970-01-01 00:00:01.299628800
568450   1970-01-01 00:00:01.331251200
568451   1970-01-01 00:00:01.329782400
568452   1970-01-01 00:00:01.331596800
568453   1970-01-01 00:00:01.338422400
Name: review_date, Length: 568454, dtype: datetime64[ns]

In [25]:
df.set_index('review_date',inplace=True)
average_review_length = df['Len_Text'].resample('M').mean()

In [26]:
average_review_length

review_date
1970-01-31    436.222083
Freq: M, Name: Len_Text, dtype: float64

#### The analysis indicates that the average length of reviews has increased over time. This suggests that users have become more detailed in their feedback, possibly providing more comprehensive information about their experiences with the products.


### 7 High-rated reviews receive more helpfulness votes:
### Do reviews with higher scores receive more helpfulness votes? Analyze the data using NumPy and Pandas

In [27]:
average_helpfulness = df.groupby('Score')['HelpfulnessNumerator'].mean().reset_index()
# Grouping By score and finding the average of the helpfulnessNumerator using mean() function and using reset_index() to convert series to DataFrame
average_helpfulness.columns=['Review_Score','Average_HelpfulnessNumerator']
# Renaming the Column names using columns

In [28]:
average_helpfulness

Review_Score  Average_HelpfulnessNumerator
0             1                      2.735211
1             2                      1.859014
2             3                      1.700962
3             4                      1.390292
4             5                      1.675228

#### This might indicate that users find very negative reviews more helpful or relevant compared to neutral or extremely positive reviews.

### 8 Products with more reviews tend to have higher average scores:
### Is there a relationship between the number of reviews a product has and its average score? Perform a statistical analysis using Pandas.

In [29]:
product_reviews = df.groupby('ProductId').agg(review_count=('Score', 'size'),average_score=('Score', 'mean')).reset_index()
product_reviews.max()

ProductId        B009WVB40S
review_count            913
average_score           5.0
dtype: object

In [30]:
correlation = product_reviews['review_count'].corr(product_reviews['average_score'])
correlation

0.0045994996649215735

#### The analysis shows that the correlation between the number of reviews and the average review score is 0.0046. This indicates an extremely weak positive correlation, suggesting that there is practically no linear relationship between these two variables. Therefore, the number of reviews a product has does not significantly impact its average review score. Further investigation into other factors might be necessary to understand what influences review scores more substantially.


### 9 Negative reviews are often shorter than positive reviews:
### How does the length of negative reviews compare to the length of positive reviews? Analyze the data using Pandas.

In [31]:
# Filtering the Reviews into Two Categories <=2 are negative and >=3 are positive
negative_reviews=df[df['Score']<=2] 
positive_reviews=df[df['Score']>=3]

In [32]:
print(negative_reviews.shape)
print(positive_reviews.shape)

(82037, 11)
(486417, 11)


In [33]:
avg_length_negative=negative_reviews['Len_Text'].mean()
avg_length_positive=positive_reviews['Len_Text'].mean()

In [34]:
print(avg_length_negative)
print(avg_length_positive)

482.6446359569463
428.3926548619806


#### The analysis shows that negative reviews are, on average, longer than positive reviews. Specifically, the average length of negative reviews is

#### 482.64 characters, while the average length of positive reviews is 428.39 characters. This suggests that users tend to provide more detailed feedback 

#### when they are dissatisfied with a product, while satisfied customers are more likely to leave brief comments expressing their satisfaction.


### 10 Review scores and helpfulness ratios differ across products</b>:
### What are the average review scores and helpfulness ratios for the top 10 products? Analyze the data using Pandas</b>.

In [35]:
df['helpfullness_ration']=df['HelpfulnessNumerator']/df['HelpfulnessDenominator']  # Finding the ration between Numerator and Denominator

In [36]:
product_stats = df.groupby('ProductId').agg(
    Max_review_score=('Score','max'),
    Average_review_score=('Score', 'mean'),
    Average_helpfulness_ratio=('helpfullness_ration', 'mean'),
    Num_of_reviews=('Score', 'size')
).reset_index()
# Grouping the DataFrame by 'ProductId' and aggregating multiple statistics for each product
# Calculate the maximum review score for each product
# Calculate the average review score for each product
# Calculate the average review score for each product
# Count the number of reviews for each product
# Reset the index to convert the grouped DataFrame back to a standard DataFrame

In [37]:
Top_10_Products=product_stats.sort_values(by='Num_of_reviews',ascending=False).head(10)
# Sorting Values based on number of reviews in desciding order and using head fuction and printing out the first 10 columns

In [38]:
Top_10_Products

ProductId  Max_review_score  Average_review_score  \
71170  B007JFMH8M                 5              4.582694   
37898  B0026RQTGE                 5              4.588608   
42257  B002QWHJOU                 5              4.588608   
42263  B002QWP89S                 5              4.588608   
42264  B002QWP8H0                 5              4.588608   
46205  B003B3OOPA                 5              4.739968   
28624  B001EO5Q64                 5              4.746032   
37798  B0026KPDG8                 5              4.310284   
20497  B000VK8AVK                 5              4.310284   
68860  B006HYLW32                 5              4.310284   

       Average_helpfulness_ratio  Num_of_reviews  
71170                   0.901042             913  
37898                   0.781684             632  
42257                   0.781684             632  
42263                   0.781684             632  
42264                   0.781684             632  
46205                   0.878274             623  
28624                   0.860282             567  
37798                   0.735122             564  
20497                   0.735122             564  
68860                   0.735122             564